lib/publiccloud/instance.pm

```diff
@@ -401,10 +400,17 @@ sub wait_for_ssh {
     my ($duration, $exit_code, $sshout, $sysout);
 
     # skip SLES4SAP as incompatible with get_public_ip
-    if (!get_var('PUBLIC_CLOUD_SLES4SAP')) {
+    if (!get_var('PUBLIC_CLOUD_SLES4SAP') && !$args{wait_stop}) {
         my $public_ip_from_provider = $self->provider->get_public_ip();
-        if ($self->public_ip ne $public_ip_from_provider) {
-            record_info('IP CHANGED', "The address we know is $self->public_ip but provider returns $public_ip_from_provider");
+        # Loop until provider returns something else than 'null' or timeout
+        while ($public_ip_from_provider =~ /null/ && ($duration = time() - $start_time) < $args{timeout}) {
```

Comment:
As you already used the word "until" in the comment why not use that directly in code. Then you also don't need the redundant comment just stating the same

```suggestion
        until ($public_ip_from_provider !~ /null/ || ($duration = time() - $start_time) >= $args{timeout}) {
```

Reply:
Please why we shall invert the logic and make the interpretation of the flow (a bit) more elaborated?